### ISSUE TITLE
Compiler mistake of types between Int and Float

### ISSUE THREAD

**Issue by gabriela-sartori** [elm-lang/error-message-catalog#71]
![image](https://cloud.githubusercontent.com/assets/6143964/11618120/edcc9892-9c7d-11e5-9830-e1362a38eef2.png)
![image](https://cloud.githubusercontent.com/assets/6143964/11618119/ea52fad0-9c7d-11e5-91a7-718b39b287eb.png)

**Comment by jvoigtlaender** [comment]
You are not showing enough code context, so it's difficult to judge whether the compiler is really at fault here. In particular, what is the type of `delay_mailbox.signal`? If it is `Signal (Float, ...)`, we might well say that the compiler is suggesting exactly the right place and thing: You might have to use `round : Float -> Int` on `sleep_time` before passing it to `ActMsg`.

**Comment by gabriela-sartori** [comment]
It's compiler fault, because it is accusing the wrong thing to be the reason of the error
delay_mailbox : Signal.Mailbox (Int, String)

**Comment by jvoigtlaender** [comment]
Another problematic example with possibly the same root came up in the issue linked above.

It is:

``` elm
test : List String -> List String
test list =
  let
    l = floor list
  in
    List.reverse list
```

and the error message is:

```
The argument to function `reverse` is causing a mismatch.
6|     List.reverse list

Function `reverse` is expecting the argument to be:

    List a

But it is:

    Float
```

So the compiler is complaining that `list` is a `Float` and is being passed to `List.reverse`, when really `list` is "obviously" a list and the actual problem is its being passed to `floor` a few lines above.

**Comment by mgold** [comment]
The error is unchanged if you omit the annotation, and in that case it's more sensible. I am _guessing_ that the compiler first infers the type of the function, and if it is well-typed, unifies that type with the annotation. One alternate approach is to, when an argument is ill-typed, see if only one of the possibilities agrees with the annotation, and if so "accuse" the other. Another is to use the annotation first rather than last. That is, before typing `floor list`, `list : a` since the compiler knows of no restrictions on its type. If the compiler starting typing with `a : List String` from the annotation, the wrong line would be the cause of the error.

Both of these approaches will probably require a fair bit of work on the compiler. Bummer.

**Comment by jvoigtlaender** [comment]
> The error is unchanged if you omit the annotation, and in that case it's more sensible. I am guessing that the compiler first infers the type of the function, and if it is well-typed, unifies that type with the annotation.

That's probably exactly what is happening. But the 0.16 announcement post paints a different picture about the helpfulness of the compiler in telling apart what was given and what was expected. So here the "natural" thing to assume is clearly that `list` is a list, because the type annotation is saying exactly that. There is hardly any more explicit evidence we could have about what is known about that argument. So it is a pity that the compiler claims the exact opposite.

**Comment by mgold** [comment]
> There is hardly any more explicit evidence we could have about what is known about that argument [than the type annotation]

Right, but sometimes annotations are wrong. If there's a simple mismatch, the error states _The type annotation for `test` does not match its definition._ I think the order reveals something about the priorities, that the annotation is expected to match the definition, not the other way around.

One benefit of the current approach is that annotating a definition does not alter the inference of its type until the last possible moment. The two cases (annotated and not) follow identical code paths through inference, simplifying the implementation, making it therefore easier to maintain and debug.

So rather than use the annotation as an input to inferencing, it might be better to use it to inform the error message for a type mismatch discovered by inference.

**Comment by jvoigtlaender** [comment]
We are doing a lot of guessing about the implementation here. And I'm not the one who would have to change the implementation, so it's cheap to "complain" (which I don't intend to do).

But, maybe for context or later reference, let me say two things from an educator's perspective:
1. The "If there's a simple mismatch, the error states _The type annotation for `test` does not match its definition_. I think the order reveals something about the priorities, that the annotation is expected to match the definition, not the other way around." is problematic.  
   I, and I'm sure others, teach the opposite when teaching typed FP. We teach: Write the type signature first. Make sure you really understand what you want as inputs and outputs, how polymorphic or not you want it, etc. Then write the code and let the compiler guide you until your code really has the type you wanted. (Of course, along the way, you might still find out that you actually wanted a slightly different type. That's fine. Revise it.)  
   But if "the annotation is expected to match the definition, not the other way around", that turns this approach on its head. With the very practical consequence that in a case like here, the outcome is bad for understanding of the learner.
2. That a case like the above can happen makes me wish to have pre-0.16 type error messages back in this regard, without the attempt at saying what is "actual" and what is "expected". Below is what Elm 0.15 said on the above `test` example. It just reported a conflict between two types, not making any judgement about which of the two types is "right" or anything. That was better than making a judgement but getting it wrong (intuitively wrong, and wrong for all learning purposes).

```
Type mismatch between the following types between lines 3 and 6:

        List String

        Float

    It is related to the following expression:

        let l = floor list in List.reverse list

Type mismatch between the following types on line 6, column 18 to 22:

        Float

        List a

    It is related to the following expression:

        list
```

**Comment by kommen** [comment]
I think https://github.com/elm-lang/error-message-catalog/pull/37 is about the same issue?

**Comment by jvoigtlaender** [comment]
Yes, it seems so. But it is less minimal than the one under discussion here. Moreover, it is not syntactically correct Elm 0.16 code.

**Comment by jvoigtlaender** [comment]
@evancz, why is this labelled as `no sscce`? We do have what I think is a minimal (and pasteable into http://elm-lang.org/try etc.) example above. The example [in this comment](https://github.com/elm-lang/error-message-catalog/issues/71#issuecomment-163032780), taken over from [here](https://github.com/elm-lang/elm-compiler/issues/1211#issuecomment-163031493).

**Comment by dan-tull** [comment]
The [earlier comment](https://github.com/elm-lang/error-message-catalog/issues/71#issuecomment-163032780) by Janis has an even more succinct example of an issue I described in [Occam's Razor for Error Messages](https://github.com/elm-lang/elm-compiler/issues/1277) so I won't repost mine here.

I will include a mention of the basic idea I presented, though. The 0.16 compiler appears to be letting the earliest type inference win and subsequent inconsistencies with that are flagged as wrong.

I think it would be better if compilation did further analysis to determine which is more likely to be incorrect based on how many other contradictions (with other expressions and type annotations) result. If there's no clear winner (both produce the same number or both produce more than some reasonable threshold of contradictions), I'd prefer a message more along the lines of the 0.15 output which at least cited all sources of inconsistency so I can more easily do further analysis myself.
